daemon/cluster/services.go

```diff
@@ -303,24 +303,56 @@ func (c *Cluster) RemoveService(input string) error {
 }
 
 // ServiceLogs collects service logs and writes them back to `config.OutStream`
-func (c *Cluster) ServiceLogs(ctx context.Context, input string, config *backend.ContainerLogsConfig, started chan struct{}) error {
+func (c *Cluster) ServiceLogs(ctx context.Context, selector *backend.LogSelector, config *backend.ContainerLogsConfig, started chan struct{}) error {
 	c.mu.RLock()
+	defer func() {
+		select {
+		case <-started:
+			// if we've started streaming logs, we are no longer holding the
+			// lock and do not have to release it
+			return
+		default:
+			// before we start, though, we're holding this lock and it needs to
+			// be released
+			c.mu.RUnlock()
+		}
+	}()
 	state := c.currentNodeState()
 	if !state.IsActiveManager() {
-		c.mu.RUnlock()
 		return c.errNoManager(state)
 	}
 
-	service, err := getService(ctx, state.controlClient, input)
-	if err != nil {
-		c.mu.RUnlock()
-		return err
+	// if ANY tasks use a TTY, don't mux streams
+	var tty bool
+	// don't rely on swarmkit to resolve IDs, do it ourselves
+	swarmSelector := &swarmapi.LogSelector{}
+	for _, s := range selector.Services {
+		service, err := getService(ctx, state.controlClient, s)
+		if err != nil {
+			return err
+		}
+		c := service.Spec.Task.GetContainer()
+		if c == nil {
+			return errors.New("logs only supported on container tasks")
+		}
+		// set TTY true if we have a TTY service, or if it's already true
+		tty = tty || c.TTY
+		swarmSelector.ServiceIDs = append(swarmSelector.ServiceIDs, service.ID)
 	}
-	container := service.Spec.Task.GetContainer()
-	if container == nil {
-		return errors.New("service logs only supported for container tasks")
+	for _, t := range selector.Tasks {
+		task, err := getTask(ctx, state.controlClient, t)
+		if err != nil {
+			return err
+		}
+		c := task.Spec.GetContainer()
+		if c == nil {
+			return errors.New("logs only supported on container tasks")
+		}
+		tty = tty || c.TTY
+		swarmSelector.TaskIDs = append(swarmSelector.TaskIDs, task.ID)
```

Comment:
The refactoring in #32154 substantially shortens this function as well, but I'm going to make this change right now anyway.